parsers/GT.py

```diff
@@ -1,163 +1,206 @@
+
+    target_datetime_utc = target_datetime.astimezone(timezone.utc)
+
+    # The API expects local (TZ) timestamps and can only return one day of data
+    # so might need to do multiple api calls if UTC day straddles multiple local days
+    target_day_utc = datetime.combine(target_datetime_utc, time(), tzinfo=timezone.utc)
+
+    target_day_utc_start = target_day_utc - timedelta(days=num_backlog_days)
+    target_day_utc_end = target_day_utc + timedelta(days=1) - timedelta(microseconds=1)
+
+    local_day_of_target_day_utc_start = datetime.combine(
+        target_day_utc_start.astimezone(TIMEZONE), time(), tzinfo=TIMEZONE
+    )
+    local_day_of_target_day_utc_end = datetime.combine(
+        target_day_utc_end.astimezone(TIMEZONE), time(), tzinfo=TIMEZONE
+    )
+    num_local_days = (
+        local_day_of_target_day_utc_end - local_day_of_target_day_utc_start
+    ).days
+
+    daily_payloads: dict[datetime, list[dict]] = {}
+    for i in range(num_local_days + 1):
+        local_day = local_day_of_target_day_utc_start + timedelta(days=i)
+        response = session.get(URL, params={"dt": local_day.strftime("%d/%m/%Y")})
+        if not response.ok:
+            raise ParserException(
+                PARSER,
+                f"Exception when fetching {target.value} error code: {response.status_code}: {response.text}",
+            )
+
+        # The JSON data returned by the API is a list of objects, each representing one technology type.
+        daily_payloads[local_day] = response.json()
+
+    if not any(daily_payload for daily_payload in daily_payloads.values()):
+        raise ParserException(
+            PARSER,
+            f"Exception when fetching {target.value}: no data available for UTC day {target_day_utc.strftime('%Y-%m-%d %Z')}. "
+            f"Note that historical data is only available for the last 1 year.",
+        )
+
+    # aggregate all individual hourly items into a dict keyed by common UTC timestamp
+    results: dict[datetime, dict] = collections.defaultdict(dict)
+    for local_day, daily_payload in daily_payloads.items():
+        for row in daily_payload:
+            # The API returns hours in the range [1, 24], so each one refers to the
+            # past hour (e.g., 1 is the time period [00:00, 01:00)). Shift the hour
+            # so each index represents the hour ahead and is in the range [0, 24),
+            # e.g., hour 0 represents the period [00:00, 01:00).
+            dt_local = local_day + timedelta(hours=int(row["hora"]) - 1)
+            dt_utc = dt_local.astimezone(timezone.utc)
+
+            # ignore data outside of range of interest
+            if dt_utc < target_day_utc_start or dt_utc > target_day_utc_end:
+                continue
+
+            # For live data, the current hour's data isn't finished updating until the hour has passed,
+            # so the current hour should not be included in the results.
+            if dt_utc >= hour_now_utc:
+                continue
```

Comment:
thought: I though of potentially still including the "live" datapoint(`dt_utc == hour_now_utc`) and we could then make it count as e.g. of being of type "forecasted"... but not sure if it would really be valid to consider that as a real forecast so I went for just skipping it

Reply:
Probably a good idea, forecasted values generally does not update and is often separate from the real data at the source as well.

We might support other types as well in the future and then a in_progress or partial might be a good thing to keep in mind.